blacktie/jatmibroker-xatmi/src/main/java/org/jboss/narayana/blacktie/jatmibroker/core/transport/hybrid/stomp/StompManagement.java

```diff
@@ -5,6 +5,7 @@
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.Socket;
+import java.nio.ByteBuffer;
```

Comment:
There are two threads intend to write into the socket when closing the service dispatcher.
http://albany.eng.hst.ams2.redhat.com/view/Narayana/job/narayana/814/PROFILE=BLACKTIE,jdk=jdk7.latest,label=linux64el6/artifact/blacktie/jatmibroker-xatmi/target/test-classes/blacktie.log
```
2015-04-29 04:39:39,986 [ConvService-Dispatcher-0]           TRACE (server.ServiceDispatcher :72  ) - Processed
2015-04-29 04:39:39,986 [ConvService-Dispatcher-0]           DEBUG (transport.Message        :48  ) - Acking message: ID:588db268-ee21-11e4-bf70-4fe43b4f0ed4
2015-04-29 04:39:39,986 [ConvService-Dispatcher-0]           TRACE (stomp.StompManagement    :70  ) - Writing on: java.net.SocketOutputStream@57c9af4d
...
2015-04-29 04:39:39,989 [main]           TRACE (server.ServiceDispatcher :116 ) - closing: ConvService-Dispatcher-0
2015-04-29 04:39:39,989 [main]           TRACE (server.ServiceDispatcher :118 ) - Closing receiver
2015-04-29 04:39:39,989 [main]           DEBUG (stomp.StompReceiverImpl  :131 ) - close
2015-04-29 04:39:39,989 [main]           DEBUG (stomp.StompReceiverImpl  :136 ) - closing socket: Socket[addr=localhost/127.0.0.1,port=61613,localport=60114]
2015-04-29 04:39:39,989 [main]           DEBUG (stomp.StompManagement    :28  ) - close
2015-04-29 04:39:39,989 [main]           TRACE (stomp.StompManagement    :70  ) - Writing on: java.net.SocketOutputStream@57c9af4d
2015-04-29 04:39:39,990 [main]           TRACE (stomp.StompManagement    :91  ) - Wrote on: java.net.SocketOutputStream@57c9af4d
2015-04-29 04:39:39,990 [main]           DEBUG (stomp.StompManagement    :35  ) - Sent disconnect
2015-04-29 04:39:39,990 [main]           TRACE (stomp.StompManagement    :96  ) - Reading from: java.net.SocketInputStream@be83fed
2015-04-29 04:39:39,990 [ConvService-Dispatcher-0]           TRACE (stomp.StompManagement    :91  ) - Wrote on: java.net.SocketOutputStream@57c9af4d
2015-04-29 04:39:39,990 [ConvService-Dispatcher-0]           DEBUG (transport.Message        :55  ) - Acked message: ID:588db268-ee21-11e4-bf70-4fe43b4f0ed4
2015-04-29 04:39:39,990 [ConvService-Dispatcher-0]           DEBUG (server.ServiceDispatcher :103 ) - Close the thread
```
It looks like the mess stomp messages sending and read from the socket is hanging as the server does not know these messages. So I agree to assemble all the data and synchronize in one write operation. Option 1 looks good and it needs to use ByteBuffer rather than StringBuffer to assemble the data. Option 2 is less code changes. Option 3 is a large change as Mike said.